File: src/main/java/dev/kensa/Kensa.java
package dev.kensa;

import dev.kensa.output.OutputStyle;
import dev.kensa.output.template.Template;
import dev.kensa.render.Renderer;
import dev.kensa.render.Renderers;
import dev.kensa.render.diagram.directive.UmlDirective;
import dev.kensa.sentence.Acronym;
import dev.kensa.sentence.Dictionary;

import java.net.MalformedURLException;
import java.net.URL;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;

public final class Kensa {

    private static final String KENSA_OUTPUT_ROOT = "kensa.output.root";
    private static final String KENSA_OUTPUT_DIR = "kensa-output";
    private static Kensa KENSA = new Kensa();

    public static Kensa configure() {
        return KENSA = new Kensa();
    }

    public static Supplier<Configuration> configuration() {
        return () -> KENSA.configuration;
    }

    private final Configuration configuration = new Configuration();

    private Kensa() {
    }

    public Kensa withIssueTrackerUrl(String url) {
        try {
            return withIssueTrackerUrl(new URL(url));
        } catch (MalformedURLException e) {
            throw new IllegalArgumentException("Invalid Issue Tracker URL specified.", e);
        }
    }

    public Kensa withIssueTrackerUrl(URL url) {
        configuration.issueTrackerUrl = url;

        return this;
    }

    public Kensa withOutputDir(String dir) {
        return withOutputDir(Paths.get(dir));
    }

    public Kensa withOutputDir(Path dir) {
        if (!dir.isAbsolute()) {
            throw new IllegalArgumentException("OutputDir must be absolute.");
        }
        configuration.outputDir = dir.endsWith(KENSA_OUTPUT_DIR) ? dir : dir.resolve(KENSA_OUTPUT_DIR);

        return this;
    }

    public <T> Kensa withRenderer(Class<T> klass, Renderer<? extends T> renderer) {
        configuration.renderers.add(klass, renderer);

        return this;
    }

    public Kensa withOutputStyle(OutputStyle outputStyle) {
        configuration.outputStyle = outputStyle;

        return this;
    }

    public Kensa withAcronyms(Acronym... acronyms) {
        configuration.dictionary.putAcronyms(acronyms);
        return this;
    }

    public Kensa withKeywords(String... keywords) {
        configuration.dictionary.putKeywords(keywords);

        return this;
    }

    public class Configuration {
        private final Dictionary dictionary;
        private Path outputDir;
        private Renderers renderers;
        private List<UmlDirective> umlDirectives;
        private OutputStyle outputStyle;
        private URL issueTrackerUrl;

        private Configuration() {
            this.outputDir = Paths.get(System.getProperty(KENSA_OUTPUT_ROOT, System.getProperty("java.io.tmpdir")), KENSA_OUTPUT_DIR);
            this.renderers = new Renderers();
            this.umlDirectives = new ArrayList<>();
            this.outputStyle = OutputStyle.MultiFile;
            this.dictionary = new Dictionary();
        }

        public Dictionary dictionary() {
            return dictionary;
        }

        public Path outputDir() {
            return outputDir;
        }

        public Renderers renderers() {
            return renderers;
        }

        public List<UmlDirective> umlDirectives() {
            return umlDirectives;
        }

        public OutputStyle outputStyle() {
            return outputStyle;
        }

        public Template createTemplate(String path, Template.Mode mode) {
            return new Template(outputDir.resolve(path), mode, issueTrackerUrl);
        }
    }
}

File: src/main/java/dev/kensa/Kensa.kt
package dev.kensa

import dev.kensa.Kensa.KENSA_OUTPUT_DIR
import dev.kensa.Kensa.KENSA_OUTPUT_ROOT
import dev.kensa.output.OutputStyle
import dev.kensa.output.template.Template
import dev.kensa.render.Renderer
import dev.kensa.render.Renderers
import dev.kensa.render.diagram.directive.UmlDirective
import dev.kensa.sentence.Acronym
import dev.kensa.sentence.Dictionary
import org.antlr.v4.runtime.atn.PredictionMode
import java.net.MalformedURLException
import java.net.URI
import java.net.URL
import java.nio.file.Path
import java.nio.file.Paths
import kotlin.reflect.KClass

object Kensa {

    internal const val KENSA_OUTPUT_ROOT = "kensa.output.root"
    internal const val KENSA_OUTPUT_DIR = "kensa-output"

    @JvmStatic
    val configuration = Configuration()

    @JvmStatic
    fun configure() = Kensa

    fun konfigure(init: Configuration.() -> Unit) {
        configuration.apply(init)
    }

    fun withIssueTrackerUrl(url: String): Kensa = apply {
        try {
            withIssueTrackerUrl(URL(url))
        } catch (e: MalformedURLException) {
            throw IllegalArgumentException("Invalid Issue Tracker URL specified.", e)
        }
    }

    fun withIssueTrackerUrl(url: URL): Kensa = apply {
        configuration.issueTrackerUrl = url
    }

    fun withOutputDir(dir: String): Kensa = withOutputDir(Paths.get(dir))

    fun withOutputDir(dir: Path): Kensa = apply {
        require(dir.isAbsolute) { "OutputDir must be absolute." }
        configuration.outputDir = if (dir.endsWith(KENSA_OUTPUT_DIR)) dir else dir.resolve(KENSA_OUTPUT_DIR)
    }

    fun <T : Any> withRenderer(klass: Class<T>, renderer: Renderer<out T>): Kensa = apply {
        configuration.renderers.add(klass, renderer)
    }

    fun <T : Any> withRenderer(kClass: KClass<T>, renderer: Renderer<out T>): Kensa = apply {
        configuration.renderers.add(kClass, renderer)
    }

    fun withOutputStyle(outputStyle: OutputStyle): Kensa = apply {
        configuration.outputStyle = outputStyle
    }

    fun withAcronyms(vararg acronyms: Acronym): Kensa = apply {
        configuration.dictionary.putAcronyms(*acronyms)
    }

    fun withKeywords(vararg keywords: String): Kensa = apply {
        configuration.dictionary.putKeywords(*keywords)
    }
}

class Configuration(
        val dictionary: Dictionary = Dictionary(),
        var outputDir: Path = Paths.get(System.getProperty(KENSA_OUTPUT_ROOT, System.getProperty("java.io.tmpdir")), KENSA_OUTPUT_DIR),
        val renderers: Renderers = Renderers(),
        var antlrPredicationMode: PredictionMode = PredictionMode.SLL,
        val umlDirectives: List<UmlDirective> = ArrayList(),
        var outputStyle: OutputStyle = OutputStyle.MultiFile,
        var issueTrackerUrl: URL = defaultIssueTrackerUrl()) {

    var keywords: Set<String> = emptySet()
        set(value) = dictionary.putKeywords(value)

    var acronyms: Set<Acronym> = emptySet()
        set(value) = dictionary.putAcronyms(value)

    fun createTemplate(path: String, mode: Template.Mode): Template = Template(outputDir.resolve(path), mode, issueTrackerUrl)

    companion object {
        private fun defaultIssueTrackerUrl(): URL = try {
            URI.create("http://empty").toURL()
        } catch (ignored: MalformedURLException) {
            throw KensaException("Unable to initialise default issue tracker url")
        }
    }
}
